drivers/timer/sam0_rtc_timer.c

```diff
@@ -0,0 +1,290 @@
+	GCLK->CLKCTRL.reg = GCLK_CLKCTRL_ID(RTC_GCLK_ID) | GCLK_CLKCTRL_CLKEN
+			    | GCLK_GEN(DT_RTC_SAM0_CLOCK_GENERATOR);
+
+	while (GCLK->STATUS.bit.SYNCBUSY) {
+		/* Synchronize GCLK. */
+	}
+
+	/* Reset module to hardware defaults. */
+	rtc_reset();
+
+	rtc_last = 0;
+
+	/* Configure RTC with 32-bit mode, configured prescaler and MATCHCLR. */
+	u16_t ctrl = RTC_MODE0_CTRL_MODE(0) | RTC_MODE0_CTRL_PRESCALER(0);
+#ifndef CONFIG_TICKLESS_KERNEL
+	ctrl |= RTC_MODE0_CTRL_MATCHCLR;
+#endif
+	rtc_sync();
+	RTC0->CTRL.reg = ctrl;
+
+	/* Enable RTC interrupt. */
+	NVIC_ClearPendingIRQ(DT_RTC_SAM0_IRQ);
+	IRQ_CONNECT(DT_RTC_SAM0_IRQ, DT_RTC_SAM0_IRQ_PRIORITY, rtc_isr, 0, 0);
+	irq_enable(DT_RTC_SAM0_IRQ);
+
+#ifdef CONFIG_TICKLESS_KERNEL
+	/* Tickless kernel lets RTC count continually and ignores overflows. */
+	RTC0->INTENSET.reg = RTC_MODE0_INTENSET_CMP0;
+#else
+	/* Non-tickless mode uses comparator together with MATCHCLR. */
+	rtc_sync();
+	RTC0->COMP[0].reg = CYCLES_PER_TICK;
+	RTC0->INTENSET.reg = RTC_MODE0_INTENSET_OVF;
+	rtc_counter = 0;
+	rtc_timeout = 0;
+#endif
+
+	/* Enable RTC module. */
+	rtc_sync();
+	RTC0->CTRL.reg |= RTC_MODE0_CTRL_ENABLE;
+
+	return 0;
+}
+
+void z_clock_set_timeout(s32_t ticks, bool idle)
+{
+	ARG_UNUSED(idle);
+
+#ifdef CONFIG_TICKLESS_KERNEL
+
+	if (ticks < 0 || ticks > MAX_TICKS) {
+		/* Handle K_FOREVER and too big ticks values. */
+		ticks = MAX_TICKS;
+	}
+
+	/* Compute number of RTC cycles until the next timeout. */
+	u32_t count = rtc_count();
+	u32_t timeout = ticks * CYCLES_PER_TICK
+			+ (count - rtc_last) % CYCLES_PER_TICK;
+
+	/* Round to the nearest tick boundary. */
+	timeout = (timeout + CYCLES_PER_TICK - 1) / CYCLES_PER_TICK
+		  * CYCLES_PER_TICK;
+
+	if (timeout < TICK_THRESHOLD) {
+		/* Trigger the interrupt right away if ticks is too low. */
+		NVIC_SetPendingIRQ(DT_RTC_SAM0_IRQ);
+		return;
+	}
+
+	rtc_sync();
+	RTC0->COMP[0].reg = count + timeout;
+
+#else /* !CONFIG_TICKLESS_KERNEL */
+
+	if (ticks < 0) {
+		/* Disable comparator for K_FOREVER and other negative
+		 * values.
+		 */
+		rtc_timeout = rtc_counter;
+		return;
+	}
+
+	if (ticks == 0) {
+		/* Trigger the interrupt right away if ticks is zero. */
+		NVIC_SetPendingIRQ(DT_RTC_SAM0_IRQ);
+		return;
+	}
+
+	/* Avoid race condition between reading counter and ISR incrementing
+	 * it.
+	 */
+	int key = irq_lock();
+
+	rtc_timeout = rtc_counter + ticks;
+	irq_unlock(key);
+
+#endif /* CONFIG_TICKLESS_KERNEL */
+}
+
+/*
+ * Gets number of ticks elapsed since the last z_clock_announce call.
+ */
+u32_t z_clock_elapsed(void)
+{
+#ifdef CONFIG_TICKLESS_KERNEL
+	return (rtc_count() - rtc_last) / CYCLES_PER_TICK;
+#else
+	return rtc_counter - rtc_last;
+#endif
+}
+
+void z_clock_idle_exit(void)
+{
+	/* Nothing to do here, RTC bookkeeping is done in ISR. */
+}
+
+u32_t _timer_cycle_get_32(void)
+{
+	return (u32_t) z_tick_get() * CYCLES_PER_TICK;
+}
+
+/*
+ * Resets RTC to its initial state.
+ */
+void rtc_reset(void)
```

Comment:
Style: should probably make this static if it's not called from elsewhere, or prefix with an appropriate name ("same_rtc_" maybe) if it will be.

Reply:
This two `rtc_*` functions are declared as static earlier in the file, but I guess there is no harm in adding the `static` keyword to their definitions. Will do.

Reply:
Oh sorry.  Then: 

Style: shouldn't write declarations for static functions.  Put the definitions in order for the compiler instead. 

More nitpicks: rtc_reset() and rtc_isr() both have one-line comments above them explaining that they reset the device and handle interrupts.  You could remove those.  The names themselves are perfectly clear.